File: src/utils/hash.ts
import { createHash } from 'node:crypto';
import { readText, writeJson, readJson, exists } from './fs.js';
import { join } from 'node:path';

export interface HashManifest {
	version: string;
	files: Record<string, string>;
}

export function hashContent(content: string): string {
	return 'sha256:' + createHash('sha256').update(content).digest('hex');
}

export async function hashFile(path: string): Promise<string> {
	const content = await readText(path);
	return hashContent(content);
}

const HASHES_FILE = '.forge/.hashes.json';

export async function readHashes(projectRoot: string): Promise<HashManifest> {
	const path = join(projectRoot, HASHES_FILE);
	if (await exists(path)) {
		return readJson<HashManifest>(path);
	}
	return { version: '0.0.0', files: {} };
}

export async function writeHashes(
	projectRoot: string,
	manifest: HashManifest,
): Promise<void> {
	await writeJson(join(projectRoot, HASHES_FILE), manifest);
}

File: internal/util/hash.go
package util

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"path/filepath"
)

// HashManifest tracks file hashes for upgrade diffing.
type HashManifest struct {
	Version string            `json:"version"`
	Files   map[string]string `json:"files"`
}

// HashContent returns a sha256 hash prefixed with "sha256:".
func HashContent(content string) string {
	h := sha256.Sum256([]byte(content))
	return fmt.Sprintf("sha256:%x", h)
}

// HashFile reads a file and returns its hash.
func HashFile(path string) (string, error) {
	content, err := ReadText(path)
	if err != nil {
		return "", err
	}
	return HashContent(content), nil
}

const hashesFile = ".forge/.hashes.json"

// ReadHashes reads the hash manifest from the project root.
func ReadHashes(projectRoot string) (HashManifest, error) {
	path := filepath.Join(projectRoot, hashesFile)
	if !Exists(path) {
		return HashManifest{Version: "0.0.0", Files: map[string]string{}}, nil
	}
	data, err := ReadText(path)
	if err != nil {
		return HashManifest{Version: "0.0.0", Files: map[string]string{}}, err
	}
	var m HashManifest
	if err := json.Unmarshal([]byte(data), &m); err != nil {
		return HashManifest{Version: "0.0.0", Files: map[string]string{}}, err
	}
	if m.Files == nil {
		m.Files = map[string]string{}
	}
	return m, nil
}

// WriteHashes writes the hash manifest to the project root.
func WriteHashes(projectRoot string, m HashManifest) error {
	return WriteJSON(filepath.Join(projectRoot, hashesFile), m)
}
